fendermint/vm/upgrade_scheduler/src/upgrade_scheduler.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::collections::BTreeMap;
+
+use anyhow::Result;
+
+/// a function type for migration
+pub type MigrationFunc = fn(i64) -> i64;
+
+/// Upgrade represents a single upgrade to be executed at a given height
+#[derive(Clone)]
+pub struct Upgrade {
+    /// the block height at which the upgrade should be executed
+    pub height: i64,
+    /// the migration function to be executed
+    pub migration: MigrationFunc,
+}
+
+/// UpgradeScheduler represents a list of upgrades to be executed at given heights
+/// During each block height we check if there is an upgrade scheduled at that
+/// height, and if so the migration for that upgrade is performed.
+#[derive(Clone)]
+pub struct UpgradeScheduler {
+    pub upgrades: BTreeMap<i64, Upgrade>,
```

Comment:
It seems redundant to index the upgrades by heights and also have a `height` property.